examples/platforms/cc2538/random.c

```diff
@@ -28,60 +28,106 @@
 
 /**
  * @file
- *   This file implements a pseudo-random number generator.
+ *   This file implements a random number generator.
  *
- * @warning
- *   This implementation is not a true random number generator and does @em satisfy the Thread requirements.
  */
 
 #include <openthread-types.h>
 
 #include <common/code_utils.hpp>
+#include <platform/radio.h>
 #include <platform/random.h>
 #include "platform-cc2538.h"
 
-static uint32_t s_state = 1;
+static void generateRandom(uint16_t aInputLength, uint8_t *aOutput, uint16_t *aOutputLength)
+{
+    uint16_t seed = 0;
+
+    HWREG(SOC_ADC_ADCCON1) &= ~(SOC_ADC_ADCCON1_RCTRL1 | SOC_ADC_ADCCON1_RCTRL0);
+    HWREG(SYS_CTRL_RCGCRFC) = SYS_CTRL_RCGCRFC_RFC0;
+
+    while (HWREG(SYS_CTRL_RCGCRFC) != SYS_CTRL_RCGCRFC_RFC0);
+
+    HWREG(RFCORE_XREG_FRMCTRL0) = RFCORE_XREG_FRMCTRL0_INFINITY_RX;
+    HWREG(RFCORE_SFR_RFST) = RFCORE_SFR_RFST_INSTR_RXON;
+
+    while (!HWREG(RFCORE_XREG_RSSISTAT) & RFCORE_XREG_RSSISTAT_RSSI_VALID);
+
+    for (uint16_t index = 0; index < aInputLength; index++)
+    {
+        if (!(index % 2))
+        {
+            seed = 0;
+
+            while (seed == 0x0000 || seed == 0x8003)
+            {
+                for (uint8_t offset = 0; offset < (8 * sizeof(seed)); offset++)
+                {
+                    seed |= (HWREG(RFCORE_XREG_RFRND) & RFCORE_XREG_RFRND_IRND);
+                    seed <<= 1;
+                }
+            }
+        }
+        else
+        {
+            seed >>= 8;
+        }
+
+        aOutput[index] = (seed & 0xff);
+    }
+
+    HWREG(RFCORE_SFR_RFST) = RFCORE_SFR_RFST_INSTR_RFOFF;
+
+    if (aOutputLength)
+    {
+        *aOutputLength = aInputLength;
+    }
+}
 
 void cc2538RandomInit(void)
 {
-    // use Extended Identifier portion of IEEE EUI-64 as the seed
-    s_state = HWREG(IEEE_EUI64 + 4);
+    uint16_t seed = 0;
+
+    generateRandom(sizeof(seed), (uint8_t *)&seed, 0);
+    HWREG(SOC_ADC_RNDL) = (seed >> 8) & 0xff;
```

Comment:
CC2538 User Guide section 16.2.2 mentions that seeding is writing to SOC_ADC_RNDL twice.